processing/src/main/java/io/druid/segment/data/BlockLayoutSerde.java

```diff
@@ -0,0 +1,309 @@
+    {
+      baseLongBuffers = GenericIndexed.read(fromBuffer, VSizeCompressedObjectStrategy.getBufferForOrder(
+          order, strategy, reader.numBytes(sizePer)
+      ));
+      this.totalSize = totalSize;
+      this.sizePer = sizePer;
+      this.order = order;
+      this.baseReader = reader;
+    }
+
+    @Override
+    public IndexedLongs get()
+    {
+      final int div = Integer.numberOfTrailingZeros(sizePer);
+      final int rem = sizePer - 1;
+      final boolean powerOf2 = sizePer == (1 << div);
+      if (powerOf2) {
+        // this provide slightly better performance than calling the LongsEncodingReader.read, probably because Java
+        // doesn't inline the method call for some reason. This should be removed when test show that performance
+        // of using read method is same as directly using accessing the longbuffer
+        if (baseReader instanceof LongsEncodingFormatSerde.LongsEncodingReader) {
+          return new BlockLayoutIndexedLongs()
+          {
+            @Override
+            public long get(int index)
+            {
+              // optimize division and remainder for powers of 2
+              final int bufferNum = index >> div;
+
+              if (bufferNum != currIndex) {
+                loadBuffer(bufferNum);
+              }
+
+              final int bufferIndex = index & rem;
+              return longBuffer.get(longBuffer.position() + bufferIndex);
+            }
+
+            protected void loadBuffer(int bufferNum)
+            {
+              CloseQuietly.close(holder);
+              holder = singleThreadedLongBuffers.get(bufferNum);
+              buffer = holder.get();
+              longBuffer = buffer.order(order).asLongBuffer();
+              currIndex = bufferNum;
+            }
+          };
+        }
+        else {
+          return new BlockLayoutIndexedLongs()
+          {
+            @Override
+            public long get(int index)
+            {
+              // optimize division and remainder for powers of 2
+              final int bufferNum = index >> div;
+
+              if (bufferNum != currIndex) {
+                loadBuffer(bufferNum);
+              }
+
+              final int bufferIndex = index & rem;
+              return reader.read(bufferIndex);
+            }
+          };
+        }
+
+      } else {
+        return new BlockLayoutIndexedLongs();
+      }
+    }
+
+    private class BlockLayoutIndexedLongs implements IndexedLongs
+    {
+      final CompressionFactory.LongEncodingFormatReader reader = baseReader.duplicate();
+      final Indexed<ResourceHolder<ByteBuffer>> singleThreadedLongBuffers = baseLongBuffers.singleThreaded();
+      int currIndex = -1;
+      ResourceHolder<ByteBuffer> holder;
+      ByteBuffer buffer;
+      LongBuffer longBuffer;
+
+      @Override
+      public int size()
+      {
+        return totalSize;
+      }
+
+      @Override
+      public long get(int index)
```

Comment:
can you just make this abstract and put all the get logic in the same general area?